lib/NativeCode/DynamicLibraryLoaderHelper/NativeRender/include/Config/Config.hpp

```diff
@@ -0,0 +1,181 @@
+#ifndef CONFIG_CLASS_HPP
+#define CONFIG_CLASS_HPP
+
+/*
+ * Copyright (c) 2024 PlayEveryWare
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#pragma once
+
+#include <filesystem>
+#include <fstream>
+#include <sstream>
+#include <array>
+#include <string>
+
+#include "json.h"
+#include "include/logging.h"
+#include "include/io_helpers.h"
+#include "include/Config/Serializable.hpp"
+#include "include/Config/Version.hpp"
+
+namespace pew::eos::config
+{
+    /**
+     * \brief Used to describe information and functionality that is common to
+     * all Config classes.
+     */
+    struct Config : public Serializable
+    {
+        // Delete the copy constructor and copy assignment operator
+        Config(const Config&) = delete;
+        Config& operator=(const Config&) = delete;
+
+        /**
+         * \brief Gets the config class and values indicated by the template
+         * parameter given.
+         * \tparam T The Config-type-derived class that is being retrieved.
+         * \return An instance of a class derived from Config that contains all
+         * values.
+         */
+        template <typename T>
+        static std::enable_if_t<std::is_base_of_v<Config, T>, std::unique_ptr<T>> get()
+        {
+            // Create the config class
+            auto config = std::unique_ptr<T>(new T());
```

Comment:
`Config` class enforces the factory pattern such that deriving objects can only be instantiated from within the `get` function. This means that private constructors are not accessible to `std::make_unique`. That means that if (in this case) you attempt to utilize `std::make_unique<T>()`, you will get the following compiler error:

`C2248: Cannot access private member declared in class.`

Using `new` with `std::unique_ptr` still retains the core benefits of `std::unique_ptr`, mainly automatic scope management that guarantees proper deletion once the `std::unique_ptr` goes out of scope or is reset, and it ensures that there is only one owner of the dynamically allocated memory. The _only_ component that is lost in this approach is that the **exception handling** that `std::make_ptr` provides is not employed.

The alternative would be to create a custom version of `std::make_unique` that does some fanciness:

```cpp
template <typename T, typename... Args>
std::unique_ptr<T> make_unique_private(Args&&... args)
{
    return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
}
```

Then utilizing it like this:

```cpp
auto config = make_unique_private<T>();
```

But that solution felt a little bit convoluted since it gets around the private constructor access limitation through clever use of templating combined with `std::forward` - and since the solution _still_ doesn't utilize `std::make_unique`, it seemed that this approach would be more difficult to maintain.

Reply:
Per conversation - this makes sense - but a comment should be added where the `new` is used to clarify why it is needed.

Reply:
Comment has been added as discussed, marking this conversation as resolved.